File: taskManagerFrontend/task-management-frontend/src/app/core/services/department.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { DepartmentRequest, DepartmentResponse, DepartmentPageResponse } from '../models/department.model';

@Injectable({
  providedIn: 'root'
})
export class DepartmentService {
  private apiUrl = `${environment.apiUrl}/departments`;

  constructor(private http: HttpClient) {}

  getAllDepartments(): Observable<DepartmentResponse[]> {
    return this.http.get<DepartmentResponse[]>(this.apiUrl);
  }

  getDepartmentsPaginated(page = 0, size = 10, sortBy = 'name', sortDir = 'asc'): Observable<DepartmentPageResponse> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sortBy', sortBy)
      .set('sortDir', sortDir);

    return this.http.get<DepartmentPageResponse>(`${this.apiUrl}/paginated`, { params });
  }

  getDepartmentById(id: number): Observable<DepartmentResponse> {
    return this.http.get<DepartmentResponse>(`${this.apiUrl}/${id}`);
  }

  createDepartment(department: DepartmentRequest): Observable<DepartmentResponse> {
    return this.http.post<DepartmentResponse>(this.apiUrl, department);
  }

  updateDepartment(id: number, department: DepartmentRequest): Observable<DepartmentResponse> {
    return this.http.put<DepartmentResponse>(`${this.apiUrl}/${id}`, department);
  }

  deleteDepartment(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: taskManager/src/main/java/com/senseiprod/taskmanager/service/DepartmentService.java
package com.senseiprod.taskmanager.service;

import com.senseiprod.taskmanager.DTOs.DepartmentDto;
import com.senseiprod.taskmanager.DTOs.PaginationDto;

import java.util.List;

public interface DepartmentService {
    DepartmentDto.Response createDepartment(DepartmentDto.Request request);
    DepartmentDto.Response getDepartmentById(Long id);
    DepartmentDto.Response updateDepartment(Long id, DepartmentDto.Request request);
    void deleteDepartment(Long id);
    List<DepartmentDto.Response> getAllDepartments();
    PaginationDto<DepartmentDto.Response> getDepartmentsPaginated(int page, int size, String sortBy, String sortDir);
}
